util/network-devp2p/src/node_record.rs

```diff
@@ -0,0 +1,62 @@
+use log::*;
+use parity_crypto::publickey::Secret;
+use crate::{disk::DiskEntity, node_table::NodeEndpoint};
+
+pub type Enr = enr::Enr<secp256k1::SecretKey>;
+
+pub struct EnrManager {
+	secret: secp256k1::SecretKey,
+	inner: Enr,
+}
+
+#[allow(dead_code)]
+impl EnrManager {
+    pub fn new(key: Secret, seq: u64) -> Option<Self> {
+		let secret = key.to_secp256k1_secret().ok()?;
+		let mut b = enr::EnrBuilder::new("v4");
+		b.seq(seq);
+		let inner = b.build(&secret).ok()?;
+		Some(Self { secret, inner })
+	}
+
+	pub fn load(key: Secret, inner: Enr) -> Option<Self> {
+		let secret = key.to_secp256k1_secret().ok()?;
+		let public = secp256k1::PublicKey::from_secret_key(&secp256k1::Secp256k1::new(), &secret);
+
+		if inner.public_key() != public {
+			warn!("ENR does not match the provided key");
+			return None;
+		}
+		Some(Self { secret, inner })
+	}
+
+	pub fn with_node_endpoint(mut self, endpoint: &NodeEndpoint) -> Self {
+		self.set_node_endpoint(endpoint);
+		self
+	}
+
+	pub fn set_node_endpoint(&mut self, endpoint: &NodeEndpoint) {
+		let seq = self.inner.seq();
+		self.inner.set_tcp_socket(endpoint.address, &self.secret).expect("Not enough data to go over the limit; qed");
+		self.inner.set_udp(endpoint.udp_port, &self.secret).expect("Not enough data to go over the limit; qed");
+		// TODO: what if we overflow here? Reset the node private key? That would require force crashing the client?
```

Comment:
Could use some help here.

Reply:
The client can be restarted, see the updater code for instance, but I'm not sure about the consequences of resetting the private key. 
But can you ELI5 what you expect to happen if `seq` is simply reset to 0?

Reply:
EIPs 778 and 868 do not tell us about the usage of `seq` but I would expect ENR consumers to use `seq` of each private key to either
* time-plot the node records or
* retain the ENR with the highest `seq` and invalidate others

In either case, resetting to 0 would break the contract defined in EIP-778 and lead to undefined behaviour in ENR consumers.

Reply:
How big is `seq`? Is this going to be a problem within our lifetimes? I'd wrap and just print an error.

Reply:
It should not be a problem unless the ENR file is manipulated. Is it OK to crash if the user provides custom ENR with seq equal to `u64::MAX`?

Reply:
Hrm, is the `sequence number` written to disk?

The `sequence number` should only be increased if the node record is modified and this seems to updated unconditionally (called at startup)?

Reply:
Sequence number is written to disk. I considered optimizing update behavior, but since geth always updates the record and increments sequence on startup, figured we may do so too.